File: cpp/questions/combinationSum.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        int cs = candidates.size();
        std::vector<std::vector<int>> ans;        
        std::vector<int> tmp;
        std::function<void(int, int)> bc = [&](int p, int tgt) -> void {
            if (tgt == 0) {
                ans.emplace_back(tmp);
                return;
            }
            for (int np=p; np<cs; ++np) {
                int nv = candidates[np];
                if (nv > tgt) {
                    continue;
                }
                tmp.push_back(nv);
                bc(np, tgt-nv);
                tmp.pop_back();
            }
        };
        bc(0, target);
        return ans;
    }
};

File: python/questions/combinationSum.py
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        lc = len(candidates)
        ans = []
        tmp = []
        def bc(p: int, tgt: int) -> None:
            if tgt == 0:
                ans.append(tmp.copy())
                return
            elif tgt < 0:
                return
            for np in range(p, lc):
                tmp.append(candidates[np])
                bc(np, tgt-candidates[np])
                tmp.pop()
        bc(0, target)
        return ans
